codec/src/test/java/io/netty/handler/codec/ByteToMessageDecoderTest.java

```diff
@@ -562,4 +565,94 @@ protected void channelRead0(ChannelHandlerContext ctx, ByteBuf msg) throws Excep
         assertNotNull(channel.pipeline().get(FixedLengthFrameDecoder.class));
         assertFalse(channel.finish());
     }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessage_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), false, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessagePartiallyReceived2x_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), true, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessage_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), false, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_MergeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.MERGE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferJustLargeEnoughToContainMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(16), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessage_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), false, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    @Test
+    public void testReuseInputBufferSufficientlyLargeToContainDuplicateMessagePartiallyReceived2x_CompositeCumulator() {
+        testReusedBuffer(Unpooled.buffer(1024), true, ByteToMessageDecoder.COMPOSITE_CUMULATOR);
+    }
+
+    private static void testReusedBuffer(ByteBuf buffer, boolean partial2, ByteToMessageDecoder.Cumulator cumulator) {
```

Comment:
Did you want the `partial2` to just be called `partial`?

Reply:
The flag is used to control whether there is a second partial component to the message or not, the `2` is meant to be for "second". I originally called it `secondPartial`, but when the build complained about this line being too long, I changed the name of this parameter instead of trying to figure out where best to insert a break in the line.

Looking at it now, I could probably make the line short enough by removing `private`, it's not needed. Assuming "`secondPartial`" is a good enough name...?

Reply:
Yeah, `secodnPartial` makes sense.